### ISSUE TITLE
seg fault like access moose.element,le,neutral

### ISSUE THREAD

**Issue by hrani** [BhallaLab/moose-core#397]
1. git commit 0e85a57b3
2. cmake info 
-- The C compiler identification is GNU 4.9.2
-- python3 (found suitable version "3.6.9", minimum required is "3.5") 
-- MOOSE Version 3.2.1.dev20200316
-- GSL version : 1.15
-- Found GSL /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so
-- Numpy is found at /opt/rh/rh-python36/root/usr/lib64/python3.6/site-packages/numpy/core/include
-- Platform: Linux-3.10.0-693.5.2.el7.x86_64-x86_64-with-centos-7.4.1708-Core
-- Build files have been written to: ~/moose-core/_build

compiles happily
3.python3
Python 3.6.9 (default, Nov 11 2019, 11:24:16) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
moose.__path__
['~/moose-core/python/moose']
>>> moose.__version__
'3.2.1.dev20200316'
>>> moose.le('/')
Segmentation fault (core dumped)

or 
import moose
>>> moose.Neutral('/r')
<moose.Neutral: id=448, dataIndex=0, path=/r[0]>
>>> moose.Neutral('/r/model')
<moose.Neutral: id=449, dataIndex=0, path=/r[0]/model[0]>
>>> moose.element('/r/model')
Segmentation fault (core dumped)

moose function like, le,element,Neutral with root is causing seg fault
But
moose.loadModel('/home/bhalla/hrani/SynPlast/AutSim/Models/synSynth7_FMRP_19Dec2019.g','/re')
  moose.reinit();moose.start(10)
but moose.element('/re') causing seg v

**Comment by dilawar** [comment]
I tried debugging it using debug build. Note the `-DCMAKE_BUILD_TYPE=Debug` option. Debug build can also be enabled by passing `-DDEBUG=ON` (option from old Makefile).

```
[dilawars@nargis _build]$ GSL_ROOT_DIR=/softwares/gsl-2.1/ cmake -DCMAKE_BUILD_TYPE=Debug ..
/usr/lib64/python3.6/distutils/dist.py:261: UserWarning: Unknown distribution option: 'extra_requires'
  warnings.warn(msg)
-- MOOSE Version 3.2.1.dev20200320
-- Building for Debug/Unit testing
-- Debug: GSL_ROOT_DIR = /softwares/gsl-2.1/
-- GSL version : 1.15
-- Found GSL /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so
-- Debug: GSL_ROOT_DIR = /softwares/gsl-2.1/
-- GSL version : 1.15
-- Found GSL /usr/lib64/libgsl.so;/usr/lib64/libgslcblas.so
-- Numpy is found at /home/bhalla/dilawars/.local/lib/python3.6/site-packages/numpy/core/include
-- Python so extension .so
-- Python module will be saved to /home/bhalla/dilawars/moose-core/pymoose/../python/moose
-- Using /usr/bin/python3-config to find Python.h
-- Python.h is found at -I/usr/include/python3.6m -I/usr/include/python3.6m
-- Platform: Linux-3.10.0-693.5.2.el7.x86_64-x86_64-with-centos-7.4.1708-Core

=======================================
If cmake did not report any error, run 
 'make' to build  MOOSE 
=======================================

-- Executable moose.bin will be at $<TARGET_FILE:moose.bin>
-- Configuring done
-- Generating done
-- Build files have been written to: /home/bhalla/dilawars/moose-core/_build
[dilawars@nargis _build]$ make -j12
.
.

// builds fine 
```

Now. Running inbuilt c++ tests.

```
$ ./moose.bin -q -u    # inbuilt c++ tests 
[dilawars@nargis _build]$ ./moose.bin -q -u
No CPU information available. Assuming single core.
on node 0, numNodes = 1, numCores = 1
[TEST  0] testAsync............................  [DONE]
[TEST  1] testMsg...  [DONE]
[TEST  2] testShell.......................  [DONE]
[TEST  3] testScheduling..  [DONE]
[TEST  4] testBuiltints....  [DONE]
[TEST  5] testKsolve...DEBUG: Gsolve.cpp:587 Grain size is 1. Num threads 1
..  [DONE]
[TEST  6] testBiophysics..  [DONE]
[TEST  7] testDiffusion........  [DONE]
[TEST  8] testHsolve...  [DONE]
[TEST  9] testMesh..........  [DONE]
[TEST 10] testSynapse.....  [DONE]
[TEST 11] testSigneur.  [DONE]
[TEST 12] testMpiMsg.  [DONE]
[TEST 13] testMpiShell................................  [DONE]
[TEST 14] testMpiBuiltins.  [DONE]
[TEST 15] testMpiScheduling  [DONE]
[TEST 16] testSchedulingProcess  [DONE]
[TEST 17] testBuiltinsProcess..  [DONE]
[TEST 18] testBiophysicsProcess...  [DONE]
[TEST 19] testSigNeurProces
```
All tests passed. No seg-fault.

I could reproduce the segv yesterday in release build. But to debug it, i had to recompile with debug symbols.

Now, the python binding.

```python
[dilawars@nargis _build]$ python3
Python 3.6.8 (default, Aug  7 2019, 17:28:10) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import moose
>>> moose.__file__
'/home/bhalla/dilawars/.local/lib/python3.6/site-packages/moose/__init__.py'
>>> moose.__version__
'3.2.1.dev20200315'
>>> moose.le('/')
Elements under /
/Msgs
/clock
/classes
/postmaster
['/Msgs', '/clock', '/classes', '/postmaster']
>>> 
```
No seg-fault. Also all python tests passes except NML and SBML since libraries are not installed. 

```
$ ctest -j12 --output-on-failure
      Start  1: moose.bin-raw-run
      Start 63: cpp_test_normal_dist
      Start 65: cpp_test_lsoda
      Start 64: cpp_test_cnpy
57/66 Test #27: core_test_moose_attribs .............................   Passed    0.71 sec
58/66 Test #65: cpp_test_lsoda ......................................   Passed    0.10 sec
59/66 Test #63: cpp_test_normal_dist ................................   Passed    0.10 sec
60/66 Test #20: core_test_vec .......................................   Passed    0.73 sec
61/66 Test #12: core_test_accessing_existing_paths ..................   Passed    0.83 sec
62/66 Test #43: core_test_Xreacs7 ...................................   Passed    0.93 sec
63/66 Test #64: cpp_test_cnpy .......................................   Passed    0.32 sec
64/66 Test #34: core_test_synchan ...................................   Passed    0.93 sec
65/66 Test  #1: moose.bin-raw-run ...................................   Passed    0.53 sec
66/66 Test #40: core_test_cylinder_diffusion_gsolve+dsolve ..........   Passed   16.47 sec

95% tests passed, 3 tests failed out of 66

Total Test time (real) =  16.48 sec

The following tests FAILED:
	 46 - support_test_sbml (Failed)
	 48 - support_test_hhfit (Failed)
	 49 - support_test_neuroml2 (Failed)
Errors while running CTest
[dilawars@nargis _build]$ 
```

See https://stackoverflow.com/questions/1762088/common-reasons-for-bugs-in-release-version-not-present-in-debug-mode 

Moreover, I could not reproduce this bug with Centos-6 where I build wheels. The gcc used there is v8.3. With version 5.x as well, this does not happen on ubuntu 14.04.  

GCC-4.9.x series is EOL (https://gcc.gnu.org/gcc-4.9/). Nargis should have an actively supported compiler. I recommend at least GCC 7.x series since I get the best performance for multithreaded Ksolve.

**Comment by dilawar** [comment]
Could reproduce with gcc5.5, both in debug and release build.

**Comment by dilawar** [comment]
With `gcc6.x`, this is not reproducible on centos-7. If you are still having trouble with it, install `devtoolset-6` which comes with gcc6. Tested on both gulgula nd nargis.

**Comment by dilawar** [comment]
More debugging:

It is less off gcc version issue and more of python3 issue. The python3.6 on centos-7 compiled with gcc-4.8.5 (see below) is likely to cause segv with python module compiled with gcc whatever version in release mode. 

```
[dilawars@gulgula _build (neuroml2)]$ python3
Python 3.6.8 (default, Aug  7 2019, 17:28:10) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

I am opening this ticket since it is likely to affect major installation. Note that the wheel works with this python as well. So either continue with this or build python3.6 with new gcc and use that this ticket is resolved.

**Comment by dilawar** [comment]
Here is a helpful messages from debugger. Adding the debug log here for future reference. 

```
   from /home1/dilawars/Work/GITLAB/moose-core/_build/python/moose/_moose.so
(gdb) l
21	 * causes problems not only for the CPython runtime, but also other
22	 * components like GNU readline.
23	 *
24	 * Accordingly, when the CLI detects it, it attempts to coerce it to a
25	 * more capable UTF-8 based alternative.
26	 *
27	 * See the documentation of the PYTHONCOERCECLOCALE setting for more details.
28	 *
29	 */
30	extern int _Py_LegacyLocaleDetected(void);
(gdb) s
Single stepping until exit from function _Z16defineAllClassesP7_object,
which has no line number information.
__cxxabiv1::__cxa_guard_acquire (g=0x7fffc213c500 <guard variable for defineAllClasses(_object*)::classes>)
    at ../../../../libstdc++-v3/libsupc++/guard.cc:232
232	  {
(gdb) n
237	    if (_GLIBCXX_GUARD_TEST_AND_ACQUIRE (g))
(gdb) 
243	    if (__gthread_active_p ())
(gdb) 
247		const int pending_bit = _GLIBCXX_GUARD_PENDING_BIT;
(gdb) 
270			 int newv = expected | waiting_bit;
(gdb) 
253		    if (__atomic_compare_exchange_n(gi, &expected, pending_bit, false,
(gdb) 
252		    int expected(0);
(gdb) 
253		    if (__atomic_compare_exchange_n(gi, &expected, pending_bit, false,
(gdb) 
333	    return acquire (g);
(gdb) 
334	  }
(gdb) 
0x00007fffc15a2cd4 in defineAllClasses(_object*) () from /home1/dilawars/Work/GITLAB/moose-core/_build/python/moose/_moose.so
(gdb) 
Single stepping until exit from function _Z16defineAllClassesP7_object,
which has no line number information.

Thread 1 "python3" received signal SIGSEGV, Segmentation fault.
0x00007fffc1df1396 in Id::eref() const () from /home1/dilawars/Work/GITLAB/moose-core/_build/python/moose/_moose.so
(gdb) 
```

**Comment by dilawar** [comment]
Ok. Found one fix or rather a workaround. MOOSE extention needs to be compiled with the same gcc version as of Python. For example, on Gulgula (same as Nargis), python3 is compiled with gcc-4.8.5 (default version).

```
[dilawars@gulgula _build (temp)]$ python3
Python 3.6.8 (default, Aug  7 2019, 17:28:10) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
```

If you compile moose with this GCC (which `cmake` would not allow currently), things work fine. I'll push some changes which allow us to use GCC-4.8.5 (needs boost libraries to replace regex and thread). I'd still call it a bug in Python3.6.8 on centos7: I should be able to compile my python module with any GCC as long as API is compatible. Details logs would be in labnotes.

More details will be in the PR which will make boost a dependency when the compiler version is less than 6.0. I can do away boost as dependencies but the multithreaded performance of solver _may_ degrade. My vote is to use BOOST freely whenever one can in this project.

@upibhalla comments?

**Comment by dilawar** [comment]
Fixed by #399 . Caused by https://github.com/BhallaLab/moose-core/pull/352 .
